src/middlewared/middlewared/plugins/acm.py

```diff
@@ -0,0 +1,736 @@
+
+    class Config:
+        namespace = 'dns.authenticator'
+        datastore = 'system.dnsauthenticator'
+
+    def __init__(self, *args, **kwargs):
+        super(DNSAuthenticatorService, self).__init__(*args, **kwargs)
+        self.schemas = DNSAuthenticatorService.get_authenticator_schemas()
+
+    @staticmethod
+    def get_authenticator_schemas():
+
+        return {
+            f_n[len('update_txt_record_'):]: [
+                Str(arg, required=True)
+                for arg in list(getattr(DNSAuthenticatorService, f_n).__code__.co_varnames)
+                [4: getattr(DNSAuthenticatorService, f_n).__code__.co_argcount]
+            ]
+            for f_n in [
+                func for func in dir(DNSAuthenticatorService)
+                if callable(getattr(DNSAuthenticatorService, func))
+                and func.startswith('update_txt_record_')
+            ]
+        }
+
+    async def common_validation(self, data, schema_name):
+        verrors = ValidationErrors()
+        if data['authenticator'] not in self.schemas:
+            verrors.add(
+                f'{schema_name}.authenticator',
+                f'FreeNAS does not support {data["authenticator"]} as an Authenticator'
+            )
+        else:
+            verrors = validate_attributes(self.schemas[data['authenticator']], data)
+
+        if verrors:
+            raise verrors
+
+    @accepts(
+        Dict(
+            'dns_authenticator_create',
+            Str('authenticator', required=True),
+            Str('name', required=True),
+            Dict('attributes', additional_attrs=True)
+        )
+    )
+    async def do_create(self, data):
+        await self.common_validation(data, 'dns_authenticator_create')
+
+        id = await self.middleware.call(
+            'datastore.insert',
+            self._config.datastore,
+            data,
+        )
+
+        return await self._get_instance(id)
+
+    @accepts(
+        Int('id', required=True),
+        Dict('attributes', additional_attrs=True, required=True)
+    )
+    async def do_update(self, id, data):
+        old = await self._get_instance(id)
+        new = old.copy()
+        new['attributes'].update(data)
+
+        await self.common_validation(data, 'dns_authenticator_update')
+
+        await self.middleware.call(
+            'datastore.update',
+            self._config.datastore,
+            id,
+            new
+        )
+
+        return await self._get_instance(id)
+
+    @accepts(
+        Int('id', required=True)
+    )
+    async def do_delete(self, id):
+
+        return await self.middleware.call(
+            'datastore.delete',
+            self._config.datastore,
+            id
+        )
+
+    @accepts(
+        Dict(
+            'update_txt_record',
+            Int('authenticator', required=True),
+            Str('key', required=True),
+            Str('domain', required=True),
+            Str('challenge', required=True)
+        )
+    )
```

Comment:
I believe it would make sense to have two args here, the first being the authentication id and the second the data, to follow CRUD convention.

You could use _get_instance(id) below avoiding all verrors code.